File: Namrata-Sanger/238. Product of Array Except Self.js
/*
Day 14
238. Product of Array Except Self
Link: https://leetcode.com/problems/product-of-array-except-self/
Level: Medium
*/

/*
 * @param {number[]} nums
 * @return {number[]}
*/

// time complexity - O(n)
// space complexity - O(1)
var productExceptSelf = function(nums) {
    let product = new Array(nums.length).fill(1);
    let mul = 1;
    
    for(let i=0; i<nums.length; i++){
        mul = mul  * nums[i];
        product[i] = mul ;
    }
    
    mul = 1;
    //calculating the product
    for(let i=nums.length-1; i>0; i--){
        product[i] = mul * product[i-1];
        mul *= nums[i];
    }
    product[0] = mul;
    return product;
};

File: Lokesh_Saana/238-product-of-array-except-self.py
# https://leetcode.com/problems/product-of-array-except-self/

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        if nums.count(0) > 1:
            return [0]*len(nums)

        product = 1
        for i in nums:
            if i != 0:
                product *= i

        if nums.count(0) == 1:
            return [product if i == nums.index(0) else 0 for i in range(len(nums))]
        return [product//i for i in nums]
